controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -331,45 +385,47 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
     }
 
     @Override
-    CompletableFuture<Void> createIndexTableIfAbsent(Data<Integer> data) {
+    CompletableFuture<Void> createHistoryIndexIfAbsent(Data<Integer> data) {
         Preconditions.checkNotNull(data);
         Preconditions.checkNotNull(data.getData());
 
+        Data<Integer> copy = new Data<>(Arrays.copyOf(data.getData(), data.getData().length), 0);
         synchronized (lock) {
-            if (indexTable == null) {
-                indexTable = new Data<>(Arrays.copyOf(data.getData(), data.getData().length), 0);
+            if (historyIndex == null) {
+                this.historyIndex = copy;
             }
         }
         return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    CompletableFuture<Data<Integer>> getIndexTable() {
+    CompletableFuture<Data<Integer>> getHistoryIndex() {
         synchronized (lock) {
-            if (this.indexTable == null) {
+            if (this.historyIndex == null) {
                 return Futures.failedFuture(StoreException.create(StoreException.Type.DATA_NOT_FOUND, getName()));
             }
-            return CompletableFuture.completedFuture(copy(indexTable));
+            return CompletableFuture.completedFuture(copy(historyIndex));
         }
     }
 
     @Override
-    CompletableFuture<Data<Integer>> getIndexTableFromStore() {
-        return getIndexTable();
+    CompletableFuture<Data<Integer>> getHistoryIndexFromStore() {
+        return getHistoryIndex();
     }
 
     @Override
-    CompletableFuture<Void> updateIndexTable(Data<Integer> updated) {
+    CompletableFuture<Void> updateHistoryIndex(Data<Integer> updated) {
         Preconditions.checkNotNull(updated);
         Preconditions.checkNotNull(updated.getData());
 
         final CompletableFuture<Void> result = new CompletableFuture<>();
+        Data<Integer> copy = updatedCopy(updated);
```

Comment:
Currently, this copy is performed under a lock, but this PR is proposing to copy outside the lock. Could you justify why this change is correct?

Reply:
We are copying the input that we receive because it is mutable and outside the control of this class. 
We dont need to copy it within a lock though as what we want to protect under lock is when we flip the reference to this new copied object.